cli/inspector.rs

```diff
@@ -4,546 +4,746 @@
 
   let inspector_map_ = inspector_map.clone();
-  let msg_handler = async move {
-    while let Some(msg) = server_msg_rx.next().await {
-      match msg {
-        ServerMsg::AddInspector(inspector_info) => {
-          let existing = inspector_map_
-            .lock()
-            .unwrap()
-            .insert(inspector_info.uuid, inspector_info);
-          if existing.is_some() {
-            panic!("UUID already in map");
-          }
-        }
-      };
-    }
-  };
+  let mut register_inspector_handler = register_inspector_rx
+    .map(|info| {
+      eprintln!(
+        "Debugger listening on {}",
+        info.get_websocket_debugger_url(&host)
+      );
+      inspector_map_
+        .lock()
+        .unwrap()
+        .insert(info.uuid, info)
+        .map(|_| panic!("Inspector UUID already in map"));
+    })
+    .collect::<()>();
+
+  let inspector_map_ = inspector_map_.clone();
+  let mut deregister_inspector_handler = future::poll_fn(|cx| {
+    inspector_map_
+      .lock()
+      .unwrap()
+      .retain(|_, info| info.canary_rx.poll_unpin(cx) == Poll::Pending);
+    Poll::<Never>::Pending
+  })
+  .fuse();
 
   let inspector_map_ = inspector_map.clone();
-  let websocket = warp::path("ws")
+  let websocket_route = warp::path("ws")
     .and(warp::path::param())
     .and(warp::ws())
-    .map(move |uuid: String, ws: warp::ws::Ws| {
-      let inspector_map__ = inspector_map_.clone();
-      ws.on_upgrade(move |socket| async move {
-        let inspector_info = {
-          if let Ok(uuid) = Uuid::parse_str(&uuid) {
-            let g = inspector_map__.lock().unwrap();
-            if let Some(inspector_info) = g.get(&uuid) {
-              inspector_info.clone()
-            } else {
-              return;
-            }
-          } else {
-            return;
-          }
-        };
-
-        // send a message back so register_worker can return...
-        let (mut ws_tx, mut ws_rx) = socket.split();
-
-        let (session_to_frontend_tx, mut session_to_frontend_rx): (
-          SessionToFrontendTx,
-          SessionToFrontendRx,
-        ) = mpsc::unbounded_channel();
-
-        // Not to be confused with the WS's uuid...
-        let session_uuid = Uuid::new_v4();
-
-        inspector_info
-          .frontend_to_inspector_tx
-          .send(FrontendToInspectorMsg::WsConnection {
-            session_to_frontend_tx,
-            session_uuid,
+    .and_then(move |uuid: String, ws: warp::ws::Ws| {
+      future::ready(
+        Uuid::parse_str(&uuid)
+          .ok()
+          .and_then(|uuid| {
+            inspector_map_
+              .lock()
+              .unwrap()
+              .get(&uuid)
+              .map(|info| info.new_websocket_tx.clone())
+              .map(|new_websocket_tx| {
+                ws.on_upgrade(move |websocket| async move {
+                  let _ = new_websocket_tx.unbounded_send(websocket);
+                })
+              })
```

Comment:
While I have no doubt this is correct, the functional style employed here is IMO gratuitous and difficult to understand when compared to the async block.